File: typescript/CaesarCipher.ts
import * as readline from 'node:readline';

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

function input(question: string): Promise<string> {
  return new Promise((resolve) => {
    rl.question(question, resolve);
  });
}

class CaesarCipher {
  static encode(plainText: string, shift: number): string {
    let result = '';
    let char = 0;

    for (let i = 0; i < plainText.length; i++) {
      const code = plainText.charCodeAt(i);

      if (code >= 97 && code <= 122) {
        char = ((code - 97) + shift) % 26 + 97;

        result += String.fromCharCode(char);
      } else {
        result += plainText[i];
      }
    }

    return result;
  }

  static decode(cipherText: string, shift: number): string {
    let result = '';
    let char = 0;

    for (let i = 0; i < cipherText.length; i++) {
      const code = cipherText.charCodeAt(i);

      if (code >= 97 && code <= 122) {
        char = ((code - 97) - shift) % 26;

        if (char < 0) char += 26;

        char += 97;

        result += String.fromCharCode(char);
      } else {
        result += cipherText[i];
      }
    }

    return result;
  }
}

async function main() {
  console.log(`CAESAR CIPHER
1.Encode
2.Decode
  `);

  const option = parseInt(await input('Choose an option: '));
  let text = await input('Enter text: ');
  const shift = parseInt(await input('Enter shift: '));

  rl.close();

  text = text.toLowerCase();

  if (option === 1) {
    const result = CaesarCipher.encode(text, shift);

    console.log(`Cipher text: ${result}`);
  }

  if (option === 2) {
    const result = CaesarCipher.decode(text, shift);

    console.log(`Plain text: ${result}`);
  }
}

main();

File: java/CaesarCipher.java
import java.util.Scanner;

public class CaesarCipher {
  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);

    System.out.println(
      """
      CAESAR CIPHER
      1. Encode
      2. Decode
      """
    );

    System.out.println("Choose an option:");

    int option = scanner.nextInt();

    scanner.nextLine();

    System.out.println("Enter text:");

    String text = scanner.nextLine();

    text = text.toLowerCase();

    System.out.println("Enter shift:");

    int shift = scanner.nextInt();

    scanner.close();

    if (option == 1) {
      String result = encode(text, shift);

      System.out.println("Cipher text: " + result);
    }

    if (option == 2) {
      String result = decode(text, shift);

      System.out.println("Plain text: " + result);
    }
  }

  static String encode(String plainText, int shift) {
    char[] result = new char[plainText.length()];
    char chr = 0;

    for (int i = 0; i < plainText.length(); i++) {
      char code = plainText.charAt(i);

      if (code >= 97 && code <= 122) {
        chr = (char)(((code - 97) + shift) % 26 + 97);

        result[i] = chr;
      } else {
        result[i] = code;
      }
    }

    return String.valueOf(result);
  }

  static String decode(String cipherText, int shift) {
    char[] result = new char[cipherText.length()];
    char chr = 0;

    for (int i = 0; i < cipherText.length(); i++) {
      char code = cipherText.charAt(i);

      if (code >= 97 && code <= 122) {
        chr = (char)(((code - 97) - shift) % 26);

        if (chr < 0) chr += 26;

        chr += 97;

        result[i] = chr;
      } else {
        result[i] = code;
      }
    }

    return String.valueOf(result);
  }
}
